### PR TITLE
feat: add new `filterPeaksRanges` function

### PR DESCRIPTION
**Pull request by jorainer**
- Add a new `filterPeaksRanges` function that allows to filter mass peaks in a `Spectra` by any (combination of) numeric spectra and peaks variables.
- Re-organize documentation of filter and subset operations (related to issue #288).

### CODE REVIEW

**Comment by jorainer** [comment]
A nice use case for this function would be to filter a LC-MS data set keeping (or removing) all mass peaks that were considered a *chromatographic peak* by *xcms*-based preprocessing. This allows e.g. to evaluate the *foreground* (i.e. all quantified ion signal) and *background* signal (i.e. background noise, or signal that was not considered a valid chromatographic peak by *xcms*) of a full data set.

I will later add such a use case to the [xcmsTutorials](https://github.com/jorainer/xcmsTutorials) tutorial/workshop.

File: R/Spectra-functions.R [hunk 1]
```diff
@@ -1182,3 +1182,184 @@ processingChunkFactor <- function(x) {
...
+#' @description
+#'
+#' The `filterPeaksRanges()` function allows to filter the peaks matrices of a
+#' [Spectra] object using any set of range-based filters on numeric spectra
+#' variables or peaks variables. These ranges can be passed to the function
+#' using the `...` as `<variable name> = <range>` pairs. `<variable name>`
+#' has to be an available spectra or peaks variable. `<range>` can be a
+#' `numeric` of length 2 defining the lower and upper boundary, or a `numeric`
+#' two-column matrix (multi-row matrices are also supported, see further
+#' below). `filterPeaksRanges(s, mz = c(200, 300))` woudl for example reduce
```
Comment (philouail) [inline R/Spectra-functions.R:1197]: would*

File: R/Spectra-functions.R [hunk 2]
```diff
@@ -1182,3 +1182,184 @@ processingChunkFactor <- function(x) {
...
+#' the peaks matrices of the `Spectra` object `s` to mass peaks with an m/z
+#' value between 200 and 300. `filterPeaksRanges()` returns the original
+#' `Spectra` object with the filter operation added to the processing queue.
+#' Thus, the filter gets only applied when the peaks data gets extracted
```
Comment (philouail) [inline R/Spectra-functions.R:1201]: maybe "**only**"  in bold to highlight it

File: R/Spectra-functions.R [hunk 3]
```diff
@@ -1182,3 +1182,184 @@ processingChunkFactor <- function(x) {
...
+#' **Filtering with multiple ranges** per spectra and peaks variables is also
+#' supported: ranges can also be provided as multi-row numeric (two-column)
+#' matrices. In this case, the above described procedure is applied for each
+#' row separately and their results are combined with a logical OR, i.e.
+#' peaks matrices are returned that match any of the conditions/filters
+#' of a row. The number of rows of the provided ranges (being it for spectra
+#' or peaks variables) have to match.
+#'
+#' **Missing value handling**: any comparison which involves a missing value
+#' (beingvit a spectra variable value, a peaks variable value or a value
```
Comment (philouail) [inline R/Spectra-functions.R:1224]: being*

File: R/Spectra-functions.R [hunk 4]
```diff
@@ -1182,3 +1182,184 @@ processingChunkFactor <- function(x) {
...
+#' in one of thevprovided ranges) is treated as a logical `FALSE`. For
```
Comment (philouail) [inline R/Spectra-functions.R:1225]: *the provided

File: R/Spectra-functions.R [hunk 5]
```diff
@@ -1182,3 +1182,184 @@ processingChunkFactor <- function(x) {
...
+#' example, if the retention time of a spectrum is `NA` and the data is
+#' filtered using a retention time range, an empty peaks matrix is returned
+#' (for `keep = TRUE`, for `keep = FALSE` the full peaks matrix is returned).
+#'
+#' @note
+#'
+#' In contrast to other *filter* functions, this function does not provide a
+#' `msLevel.` parameter to apply the filter only on spectra of the specified
```
Comment (philouail) [inline R/Spectra-functions.R:1233]: extra point at `msLevel`

File: R/Spectra-functions.R [hunk 6]
```diff
@@ -1182,3 +1182,184 @@ processingChunkFactor <- function(x) {
...
+#' MS levels. In contrast, to apply no, or different, filters to spectra from
```
Comment (philouail) [inline R/Spectra-functions.R:1234]: mmm i'm confused by this sentence

File: R/Spectra.R [hunk 7]
```diff
@@ -331,31 +331,23 @@ NULL
...
-#' @section Data subsetting, filtering and merging:
+#' @section Filter spectra data:
 #'
-#' Subsetting and filtering of `Spectra` objects can be performed with the below
-#' listed methods.
+#' Filter a `Spectra` object based on the spectra data. This includes subset
+#' operations that reduce the number of spectra in the object as well as filters
+#' that reduce the *content* of the `Spectra` object. See section
+#' *Filter peaks data* below for functions that filter the peaks data of a
+#' `Spectra`.
```
Comment (philouail) [inline R/Spectra.R:340]: maybe to make it extremely clear, we can say the number of spectra in the Spectra object will not change with filters until applyProcessing() is run ?

Comment (philouail) [reply R/Spectra.R:340]: I guess here or at the beginning of the filter section

File: vignettes/Spectra.Rmd [hunk 8]
```diff
@@ -360,25 +360,39 @@ Similar to spectra variables it is also possible to replace values for
...
+*spectra data* and operations that filter *mass peak data*. A third category of
+function allows to aggregate data within a `Spectra` or to merge and combine
+multiple `Spectra` objects into one. Functions of the various categories are
+described in the following subsections. Please refer to the function's
+documentation for more details and information.
 
+
+### Filter spectra data
+
+These functions comprise subset operations that reduce the total number of
```
Comment (philouail) [inline vignettes/Spectra.Rmd:376]: Maybe add "directly" or immediately" ?

File: vignettes/Spectra.Rmd [hunk 9]
```diff
@@ -390,53 +404,59 @@ affecting the total number of spectra are (in alphabetic order):
...
+- `selectSpectraVariables()`: reduces the (spectra) data within the object to
+  the selected spectra variables.
+
 
-Filter functions that return the same number of spectra, but affect/subset the
-peaks data (m/z and intensity values) within each spectrum are:
+### Filter or aggregate mass peak data
+
+These function filter or aggregate the mass peak data (`peaksData()`) of each
+spectrum in a `Spectra` without changing the total number of spectra.
```
Comment (philouail) [inline vignettes/Spectra.Rmd:422]: as before I would precise "until the applyProcessing() function in run"

**Review by philouail — APPROVED** [review]
The codes looks all good to me. 

Thanks for reorganizing the documentation I believe it is much more clear now. I put some comments to signals typos and maybe some suggestions. But i think it's great !

**Comment by jorainer** [comment]
Thanks @philouail ! I addressed all of your points! Hope documentation is now a bit clearer. Happy for additional/extra feedback, otherwise I proceed and merge tomorrow.

**Pull request merged**